addons/zylann.hterrain/hterrain_detail_layer.gd

```diff
@@ -304,116 +302,84 @@ func _notification(what: int):
 	if _multimesh_need_regen:
 		_regen_multimesh()
 		_multimesh_need_regen = false
 		# Crash workaround for Godot 3.1
 		# See https://github.com/godotengine/godot/issues/32500
-		for k in _chunks:
-			var mmi = _chunks[k]
+		for k in _mmis:
+			var mmi = _mmis[k]
 			mmi.set_multimesh(_multimesh)
-
-	var local_viewer_pos = terrain.global_transform.affine_inverse() * viewer_pos
-
-	var viewer_cx = local_viewer_pos.x / CHUNK_SIZE
-	var viewer_cz = local_viewer_pos.z / CHUNK_SIZE
-
-	var cr = int(view_distance) / CHUNK_SIZE + 1
-
-	var cmin_x = viewer_cx - cr
-	var cmin_z = viewer_cz - cr
-	var cmax_x = viewer_cx + cr
-	var cmax_z = viewer_cz + cr
-
-	var map_res = terrain.get_data().get_resolution()
-	var map_scale = terrain.map_scale
-
-	var terrain_size_x = map_res * map_scale.x
-	var terrain_size_z = map_res * map_scale.z
-
-	var terrain_chunks_x = terrain_size_x / CHUNK_SIZE
-	var terrain_chunks_z = terrain_size_z / CHUNK_SIZE
-
-	if cmin_x < 0:
-		cmin_x = 0
-	if cmin_z < 0:
-		cmin_z = 0
-	if cmax_x > terrain_chunks_x:
-		cmax_x = terrain_chunks_x
-	if cmax_z > terrain_chunks_z:
-		cmax_z = terrain_chunks_z
-
-	if DEBUG and visible:
+			
+	if DEBUG:
 		_debug_cubes.clear()
-		for cz in range(cmin_z, cmax_z):
-			for cx in range(cmin_x, cmax_x):
-				_add_debug_cube(terrain, _get_chunk_aabb(terrain, Vector3(cx, 0, cz) * CHUNK_SIZE))
-
+	
 	var terrain_transform : Transform = terrain.get_internal_transform()
-
-	for cz in range(cmin_z, cmax_z):
-		for cx in range(cmin_x, cmax_x):
-
+	var terrain_scale = terrain.map_scale
+	
+	for cz in range(cmin.y, cmax.y):
+		for cx in range(cmin.x, cmax.x):
+			var x_pos = int((cx * _chunk_size) / terrain_scale.x)
+			var z_pos = int((cz * _chunk_size) / terrain_scale.z)
+			
+			if DEBUG:
+				_add_debug_cube(terrain, aabbs[Vector2(x_pos, z_pos)])
+				
 			var cpos2d = Vector2(cx, cz)
-			if _chunks.has(cpos2d):
+			if _mmis.has(cpos2d):
 				continue
 
-			var aabb = _get_chunk_aabb(terrain, Vector3(cx, 0, cz) * CHUNK_SIZE)
-			var d = (aabb.position + 0.5 * aabb.size).distance_to(local_viewer_pos)
-
-			if d < view_distance:
-				_load_chunk(terrain_transform, cx, cz, aabb)
+			var aabb = aabbs[Vector2(x_pos, z_pos)]
+			var sqr_dist = (aabb.position + 0.5 * aabb.size).distance_squared_to(local_viewer_pos)
+			if sqr_dist < view_distance * view_distance:
+				_load_mmi(terrain_transform, cx, cz, aabb)
 
 	var to_recycle = []
 
-	for k in _chunks:
-		var chunk = _chunks[k]
-		var aabb = _get_chunk_aabb(terrain, Vector3(k.x, 0, k.y) * CHUNK_SIZE)
-		var d = (aabb.position + 0.5 * aabb.size).distance_to(local_viewer_pos)
-		if d > view_distance:
+	for k in _mmis:
+		var mmi = _mmis[k]
+		var x_pos = int((k.x * _chunk_size) / terrain_scale.x)
+		var z_pos = int((k.y * _chunk_size) / terrain_scale.z)
+		var aabb = aabbs[Vector2(x_pos, z_pos)]
+		var sqr_dist = (aabb.position + 0.5 * aabb.size).distance_squared_to(local_viewer_pos)
+		if  sqr_dist > view_distance * view_distance:
```

Comment:
Same here about conversions

Reply:
Same as above. I've updated the code to use Vector2(cx, cz) to lookup into the dictionary